Signal/src/network/GiphyDownloader.swift

```diff
@@ -0,0 +1,410 @@
+    }
+
+    private func clearCallbacks() {
+        AssertIsOnMainThread()
+
+        success = nil
+        failure = nil
+    }
+
+    public func requestDidSucceed(asset: GiphyAsset) {
+        AssertIsOnMainThread()
+
+        success?(self, asset)
+
+        // Only one of the callbacks should be called, and only once.
+        clearCallbacks()
+    }
+
+    public func requestDidFail() {
+        AssertIsOnMainThread()
+
+        failure?(self)
+
+        // Only one of the callbacks should be called, and only once.
+        clearCallbacks()
+    }
+}
+
+// Represents a downloaded gif asset.
+//
+// The blob on disk is cleaned up when this instance is deallocated,
+// so consumers of this resource should retain a strong reference to
+// this instance as long as they are using the asset.
+@objc class GiphyAsset: NSObject {
+    static let TAG = "[GiphyAsset]"
+
+    let rendition: GiphyRendition
+    let filePath: String
+
+    init(rendition: GiphyRendition,
+         filePath: String) {
+        self.rendition = rendition
+        self.filePath = filePath
+    }
+
+    deinit {
+        // Clean up on the asset on disk.
+        let filePathCopy = filePath
+        DispatchQueue.global().async {
+            do {
+                let fileManager = FileManager.default
+                try fileManager.removeItem(atPath:filePathCopy)
+            } catch let error as NSError {
+                owsFail("\(GiphyAsset.TAG) file cleanup failed: \(filePathCopy), \(error)")
+            }
+        }
+    }
+}
+
+// A simple LRU cache bounded by the number of entries.
+class LRUCache<KeyType: Hashable & Equatable, ValueType> {
+
+    private var cacheMap = [KeyType: ValueType]()
+    private var cacheOrder = [KeyType]()
+    private let maxSize: Int
+
+    init(maxSize: Int) {
+        self.maxSize = maxSize
+    }
+
+    public func get(key: KeyType) -> ValueType? {
+        guard let value = cacheMap[key] else {
+            return nil
+        }
+
+        // Update cache order.
+        cacheOrder = cacheOrder.filter { $0 != key }
+        cacheOrder.append(key)
+
+        return value
+    }
+
+    public func set(key: KeyType, value: ValueType) {
+        cacheMap[key] = value
+
+        // Update cache order.
+        cacheOrder = cacheOrder.filter { $0 != key }
+        cacheOrder.append(key)
+
+        while cacheOrder.count > maxSize {
+            guard let staleKey = cacheOrder.first else {
+                owsFail("Cache ordering unexpectedly empty")
+                return
+            }
+            cacheOrder.removeFirst()
+            cacheMap.removeValue(forKey:staleKey)
+        }
+    }
+}
+
+private var URLSessionTaskGiphyAssetRequest: UInt8 = 0
+
+// This extension is used to punch an asset request onto a download task.
+extension URLSessionTask {
+    var assetRequest: GiphyAssetRequest {
+        get {
+            return objc_getAssociatedObject(self, &URLSessionTaskGiphyAssetRequest) as! GiphyAssetRequest
```

Comment:
should this be optional (`as?`)? Or do we want to crash when  an `assetRequest` is nil?

Reply:
Yeah, this should always be non-nil, its set when the task is created and before its started with `resume`.